File: sdk/@launchdarkly/observability-android/lib/src/main/kotlin/com/launchdarkly/observability/replay/masking/MaskApplier.kt
package com.launchdarkly.observability.replay.masking

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.PointF
import android.graphics.Rect
import androidx.core.graphics.withScale
import kotlin.math.abs
import kotlin.math.max

class MaskApplier {
    // Matches the iOS/Flutter renderer:
    //   hull  = white 0.50 -> round(0.50 * 255) = 128 = 0x80 (area the mask swept)
    //   before = white 0.52 -> round(0.52 * 255) = 133 = 0x85 (precise position)
    private val hullMaskPaint = Paint().apply {
        color = Color.rgb(0x80, 0x80, 0x80)
        style = Paint.Style.FILL
    }
    private val beforeMaskPaint = Paint().apply {
        color = Color.rgb(0x85, 0x85, 0x85)
        style = Paint.Style.FILL
    }

    private val maskIntRect = Rect()

    fun drawMasks(canvas: Canvas, maskPairsList: List<Pair<Mask, Mask?>>, scaleFactor: Float) {
        if (maskPairsList.isEmpty()) return

        canvas.withScale(scaleFactor, scaleFactor) {
            val path = Path()
            maskPairsList.forEach { pairOfMasks ->
                drawMask(pairOfMasks, path, this)
            }
        }
    }

    private fun drawMask(pairOfMasks: Pair<Mask, Mask?>, path: Path, canvas: Canvas) {
        val (before, after) = pairOfMasks
        // When the mask drifted between the before/after passes, fill the convex
        // hull spanning both positions first so the strip it swept across is
        // covered; then always stamp the precise "before" position on top.
        if (after != null) {
            drawHull(before, after, path, canvas, hullMaskPaint)
        }
        drawMask(before, path, canvas, beforeMaskPaint)
    }

    private fun drawMask(mask: Mask, path: Path, canvas: Canvas, paint: Paint) {
        if (mask.points != null) {
            val pts = mask.points

            path.reset()
            path.moveTo(pts[0], pts[1])
            path.lineTo(pts[2], pts[3])
            path.lineTo(pts[4], pts[5])
            path.lineTo(pts[6], pts[7])
            path.close()

            canvas.drawPath(path, paint)
        } else {
            maskIntRect.left = mask.rect.left.toInt()
            maskIntRect.top = mask.rect.top.toInt()
            maskIntRect.right = mask.rect.right.toInt()
            maskIntRect.bottom =  mask.rect.bottom.toInt()
            canvas.drawRect(maskIntRect, paint)
        }
    }

    /** Fills the convex hull of the two masks' eight corners. Mirrors the iOS/Flutter `drawHull`. */
    private fun drawHull(before: Mask, after: Mask, path: Path, canvas: Canvas, paint: Paint) {
        val corners = ArrayList<PointF>(8)
        addCorners(before, corners)
        addCorners(after, corners)
        val hull = convexHull8(corners)
        if (hull.size < 3) return

        path.reset()
        path.moveTo(hull[0].x, hull[0].y)
        for (i in 1 until hull.size) {
            path.lineTo(hull[i].x, hull[i].y)
        }
        path.close()
        canvas.drawPath(path, paint)
    }

    private fun addCorners(mask: Mask, out: MutableList<PointF>) {
        val pts = mask.points
        if (pts != null) {
            out.add(PointF(pts[0], pts[1]))
            out.add(PointF(pts[2], pts[3]))
            out.add(PointF(pts[4], pts[5]))
            out.add(PointF(pts[6], pts[7]))
        } else {
            val r = mask.rect
            out.add(PointF(r.left, r.top))
            out.add(PointF(r.right, r.top))
            out.add(PointF(r.right, r.bottom))
            out.add(PointF(r.left, r.bottom))
        }
    }

    fun mergeMasksMap(
        beforeMasksMap: List<List<Mask>?>,
        afterMasksMap: List<List<Mask>?>
    ): List<List<Pair<Mask, Mask?>>?>? {
        if (afterMasksMap.count() != beforeMasksMap.count()) {
            return null
        }

        val result = buildList(beforeMasksMap.size) {
            for (i in beforeMasksMap.indices) {
                val before = beforeMasksMap[i]
                val after = afterMasksMap[i]
                if (before == null) {
                    if (after == null) {
                        add(null)
                        continue
                    } else return null
                }
                if (after != null) {
                    val merged = mergeMasks(before, after) ?: return null
                    add(merged)
                } else {
                    return null
                }
            }
        }

        return result
    }

    /**
     * Reconciles the "before" and "after" mask passes. A mask that barely moved
     * is paired with a `null` "after" (the before already covers it); a mask that
     * moved is kept as a pair so [drawMask] spans both positions with a convex
     * hull. Returns `null` (drop the frame) when the passes don't line up — counts
     * or view ids differ, or a mask drifted further than its own size so the gap
     * can't be safely covered. Mirrors the iOS/Flutter `MaskStabilizer`.
     */
    private fun mergeMasks(
        beforeMasks: List<Mask>,
        afterMasks: List<Mask>
    ): List<Pair<Mask, Mask?>>? {
        if (afterMasks.count() != beforeMasks.count()) {
            return null
        }

        if (afterMasks.isEmpty()) {
            return listOf()
        }

        val resultMasks = mutableListOf<Pair<Mask, Mask?>>()
        for ((before, after) in beforeMasks.zip(afterMasks)) {
            if (before.viewId != after.viewId) {
                return null
            }
            val dy = abs(after.rect.top - before.rect.top)
            val dx = abs(after.rect.left - before.rect.left)
            val dRight = abs(after.rect.right - before.rect.right)
            val dBottom = abs(after.rect.bottom - before.rect.bottom)

            if (max(max(dx, dy), max(dRight, dBottom)) <= MOVE_TOLERANCE) {
                // Both position and size are within tolerance; "before" already
                // covers the area. Checking all four edges (not just the
                // top-left) ensures a mask that grows in place — without its
                // top-left moving — still gets hull-drawn below instead of being
                // dropped, which would expose the newly grown region.
                resultMasks += Pair(before, null)
                continue
            }

            val coversX = dx * OVERLAP_TOLERANCE < before.rect.width() - MOVE_TOLERANCE
            val coversY = dy * OVERLAP_TOLERANCE < before.rect.height() - MOVE_TOLERANCE
            if (!coversX || !coversY) {
                // Moved further than its own size; the gap can't be safely covered.
                return null
            }

            resultMasks += Pair(before, after)
        }

        return resultMasks
    }

    private companion object {
        /** Movement under this many pixels on either axis is treated as the same position. */
        const val MOVE_TOLERANCE = 1f

        /** Slack required between the observed delta and the mask's own size before covering it with a hull. */
        const val OVERLAP_TOLERANCE = 1.1f

        /** Z-component of the cross product of OA x OB; negative for a clockwise turn (y-down). */
        private fun cross(o: PointF, a: PointF, b: PointF): Float =
            (a.x - o.x) * (b.y - o.y) - (a.y - o.y) * (b.x - o.x)

        /**
         * Convex hull of a small point cloud (~8 points) via gift-wrapping, matching
         * the iOS/Flutter `convexHull8`. Returns the input unchanged for fewer than
         * four points.
         */
        fun convexHull8(points: List<PointF>): List<PointF> {
            if (points.size < 4) {
                return points
            }

            val hull = ArrayList<PointF>(points.size)

            var startPoint = points[0]
            for (point in points) {
                if (point.x < startPoint.x) {
                    startPoint = point
                }
            }

            var currentPoint = startPoint
            do {
                hull.add(currentPoint)
                var nextPoint = points[0]

                for (candidate in points) {
                    if (nextPoint == currentPoint) {
                        nextPoint = candidate
                        continue
                    }
                    if (cross(currentPoint, candidate, nextPoint) < 0) {
                        nextPoint = candidate
                    }
                }
                currentPoint = nextPoint

                // Safety valve against a non-terminating wrap from coincident points.
                if (hull.size > points.size) {
                    break
                }
            } while (currentPoint != startPoint)

            return hull
        }
    }
}

File: sdk/@launchdarkly/flutter/packages/observability/lib/src/platform/io/masking/mask_applier.dart
import 'dart:ui' as ui;

import 'package:flutter/rendering.dart';

import 'mask_operation.dart';

/// Paints the resolved [MaskPair]s onto the captured frame, producing a new
/// redacted image.
///
/// Flutter counterpart of the native `MaskApplier`. For each pair:
///   - when the mask drifted between the before/after passes ([MaskPair.after]
///     is non-null), the convex hull spanning *both* positions is filled in the
///     standard grey, so the strip the mask swept across is fully covered; and
///   - the precise "before" quad is always drawn on top in a slightly different
///     grey, mirroring the native renderer so a merged/animated region is
///     visibly distinguishable when reviewing a recording.
///
/// Operations carry *logical* geometry; the applier scales by [pixelRatio] to
/// land on the captured (physical-resolution) image.
class MaskApplier {
  const MaskApplier();

  // Mirrors the native `MaskApplier`:
  //   standard  = UIColor(white: 0.50) -> round(0.50 * 255) = 128 = 0x80
  //   duplicate = UIColor(white: 0.52) -> round(0.52 * 255) = 133 = 0x85
  static const ui.Color _standardMaskColor = ui.Color(0xFF808080);
  static const ui.Color _duplicateMaskColor = ui.Color(0xFF858585);
  static const double _cornerRadius = 2.0;

  /// Draws [pairs] over [image] and returns the redacted image. The caller owns
  /// both [image] and the returned image and must dispose them.
  Future<ui.Image> apply(
    ui.Image image,
    List<MaskPair> pairs,
    double pixelRatio,
  ) async {
    final recorder = ui.PictureRecorder();
    final canvas = ui.Canvas(recorder);

    canvas.drawImage(image, ui.Offset.zero, ui.Paint());

    final fillPaint = ui.Paint()..color = _standardMaskColor;
    final duplicatePaint = ui.Paint()..color = _duplicateMaskColor;

    // Scales logical geometry into the captured image's physical pixels.
    final scale = Matrix4.diagonal3Values(pixelRatio, pixelRatio, 1);

    for (final pair in pairs) {
      final after = pair.after;
      if (after != null) {
        // Merge the two transformed positions via their convex hull so the
        // area the mask travelled across between the passes is covered.
        _drawHull(canvas, pair.before.quad, after.quad, pixelRatio, fillPaint);
      }
      // Always stamp the precise "before" position on top. An affine mask is a
      // rounded rect transformed into place (matching the native `drawRect`); a
      // perspective-projected (3D) mask is drawn as its raw quad (matching the
      // native `drawQuad`).
      if (pair.before.isAffine) {
        _drawRect(canvas, pair.before, scale, duplicatePaint);
      } else {
        _drawQuad(canvas, pair.before.quad, pixelRatio, duplicatePaint);
      }
    }

    final picture = recorder.endRecording();
    try {
      return await picture.toImage(image.width, image.height);
    } finally {
      picture.dispose();
    }
  }

  /// Fills the rounded-rect mask transformed into physical pixels. The corners
  /// are rounded in the mask's *local* space (radius [_cornerRadius]) before the
  /// affine transform is applied, matching the native `drawRect`.
  void _drawRect(
    ui.Canvas canvas,
    MaskOperation operation,
    Matrix4 scale,
    ui.Paint paint,
  ) {
    final rrect = ui.RRect.fromRectXY(
      operation.localRect,
      _cornerRadius,
      _cornerRadius,
    );
    final localPath = ui.Path()..addRRect(rrect);
    final matrix = scale.multiplied(operation.transform);
    canvas.drawPath(localPath.transform(matrix.storage), paint);
  }

  /// Fills the quad polygon, scaled into physical pixels. Used for masks whose
  /// transform has a perspective component, so the projection isn't a
  /// parallelogram a single rounded rect could represent. Mirrors the native
  /// `drawQuad`.
  void _drawQuad(
    ui.Canvas canvas,
    Quad quad,
    double pixelRatio,
    ui.Paint paint,
  ) {
    final points = quad.points;
    final path = ui.Path()
      ..moveTo(points[0].dx * pixelRatio, points[0].dy * pixelRatio);
    for (var i = 1; i < points.length; i++) {
      path.lineTo(points[i].dx * pixelRatio, points[i].dy * pixelRatio);
    }
    path.close();
    canvas.drawPath(path, paint);
  }

  /// Fills the convex hull of the two quads (8 corner points), scaled into
  /// physical pixels. Mirrors the native `drawHull`.
  void _drawHull(
    ui.Canvas canvas,
    Quad quad1,
    Quad quad2,
    double pixelRatio,
    ui.Paint paint,
  ) {
    final hull = convexHull8(<ui.Offset>[...quad1.points, ...quad2.points]);
    if (hull.length < 3) {
      return;
    }
    final path = ui.Path()
      ..moveTo(hull[0].dx * pixelRatio, hull[0].dy * pixelRatio);
    for (var i = 1; i < hull.length; i++) {
      path.lineTo(hull[i].dx * pixelRatio, hull[i].dy * pixelRatio);
    }
    path.close();
    canvas.drawPath(path, paint);
  }
}

/// Z-component of the cross product of `OA` x `OB`. Negative when `O -> A -> B`
/// makes a clockwise turn in Flutter's y-down coordinate space.
double _cross(ui.Offset o, ui.Offset a, ui.Offset b) =>
    (a.dx - o.dx) * (b.dy - o.dy) - (a.dy - o.dy) * (b.dx - o.dx);

/// Convex hull of a small point cloud (~8 points) via a gift-wrapping search,
/// matching the native `convexHull8`. Returns the input unchanged for fewer than
/// four points.
///
/// Used to wrap the eight corners of a before/after mask pair into a single
/// polygon that covers the strip the mask travelled across.
List<ui.Offset> convexHull8(List<ui.Offset> points) {
  if (points.length < 4) {
    return points;
  }

  final hull = <ui.Offset>[];

  var startPoint = points[0];
  for (final point in points) {
    if (point.dx < startPoint.dx) {
      startPoint = point;
    }
  }

  var currentPoint = startPoint;
  do {
    hull.add(currentPoint);
    var nextPoint = points[0];

    for (final candidate in points) {
      if (nextPoint == currentPoint) {
        nextPoint = candidate;
        continue;
      }
      if (_cross(currentPoint, candidate, nextPoint) < 0) {
        nextPoint = candidate;
      }
    }
    currentPoint = nextPoint;

    // Safety valve against a non-terminating wrap from degenerate / coincident
    // points (the native version relies on exact CGPoint equality; floating
    // point input could otherwise loop).
    if (hull.length > points.length) {
      break;
    }
  } while (currentPoint != startPoint);

  return hull;
}
